src/lib.rs

```diff
@@ -0,0 +1,238 @@
+use crate::state::State;
+use fil_actors_runtime::runtime::{ActorCode, Runtime};
+use fil_actors_runtime::{actor_error, cbor, ActorDowncast, ActorError, INIT_ACTOR_ADDR};
+use fvm_ipld_blockstore::Blockstore;
+use fvm_ipld_encoding::RawBytes;
+use fvm_ipld_hamt::BytesKey;
+use fvm_shared::econ::TokenAmount;
+use fvm_shared::error::ExitCode;
+use fvm_shared::{MethodNum, METHOD_CONSTRUCTOR};
+use ipc_gateway::{ApplyMsgParams, CrossMsg, IPCAddress, StorableMsg};
+use num_derive::FromPrimitive;
+use num_traits::{FromPrimitive, Zero};
+
+pub use crate::types::{ConstructorParams, PreCommitParams, RevokeParams};
+
+mod atomic;
+mod state;
+mod types;
+
+fil_actors_runtime::wasm_trampoline!(Actor);
+
+/// Atomic execution coordinator actor methods available
+#[derive(FromPrimitive)]
+#[repr(u64)]
+pub enum Method {
+    Constructor = METHOD_CONSTRUCTOR,
+    PreCommit = 2,
+    Revoke = 3,
+}
+
+/// Atomic execution coordinator actor
+pub struct Actor;
+
+impl Actor {
+    fn constructor<BS, RT>(rt: &mut RT, params: ConstructorParams) -> Result<(), ActorError>
+    where
+        BS: Blockstore,
+        RT: Runtime<BS>,
+    {
+        rt.validate_immediate_caller_is(std::iter::once(&*INIT_ACTOR_ADDR))?;
+
+        let st = State::new(rt.store(), params).map_err(|e| {
+            e.downcast_default(ExitCode::USR_ILLEGAL_STATE, "Failed to create actor state")
+        })?;
+        rt.create(&st)?;
+        Ok(())
+    }
+
+    /// Records a pre-commitment from an actor to perform an atomic
+    /// execution. This method is to be invoked by a wrapped crossnet
+    /// message originating in one of the execution actors involved in
+    /// the atomic execution. Once the coordinator actor collects
+    /// pre-commitments from all the execution actors, it emits for
+    /// each of the execution actors a crossnet message triggering the
+    /// specified method to commit the atomic execution.
+    fn pre_commit<BS, RT>(rt: &mut RT, params: ApplyMsgParams) -> Result<bool, ActorError>
+    where
+        BS: Blockstore,
+        RT: Runtime<BS>,
+    {
+        let st: State = rt.state()?;
+        rt.validate_immediate_caller_is(std::iter::once(&st.ipc_gateway_address))?;
```

Comment:
We expect the `ipc_address` to be the caller as these messages are expected to be received as cross-net messages, right? (can we add a slight comment here so it is clear? Thanks).

Reply:
This is mentioned in the doc comment of the method:

> This method is to be invoked by a wrapped crossnet message

But I agree, it's better to add a bit of clarification in the code, as well.

Reply:
Thanks!